autopush/db.py

```diff
@@ -432,45 +445,70 @@ def generate_last_connect_values(date):
-    while True:
-        if start_table:  # pragma nocover
-            result = client.list_tables(ExclusiveStartTableName=start_table)
-        else:
-            result = client.list_tables()
-        for table in result.get('TableNames', []):
-            yield table
-        start_table = result.get('LastEvaluatedTableName', None)
-        if not start_table:
-            break
+def table_exists(tablename, boto_resource=None):
+    # type: (str, DynamoDBResource) -> bool
+    """Determine if the specified Table exists"""
+    try:
+        return boto_resource.Table(tablename).table_status in [
+            'CREATING', 'UPDATING', 'ACTIVE']
+    except ClientError:
+        return False
 
 
-def table_exists(tablename, client=None):
-    """Determine if the specified Table exists"""
-    if not client:
-        client = g_client
-    return tablename in list_tables(client)
+class DynamoDBResource(threading.local):
+    def __init__(self, **kwargs):
+        conf = kwargs
+        if not conf.get("endpoint_url"):
+            conf["endpoint_url"] = os.getenv("AWS_LOCAL_DYNAMODB")
+        # If there is no endpoint URL, we must delete the entry
+        if "endpoint_url" in conf and not conf["endpoint_url"]:
+            del(conf["endpoint_url"])
+        region = conf.get("region_name",
+                          os.getenv("AWS_REGION_NAME", "us-east-1"))
+        if "region_name" in conf:
+            del(conf["region_name"])
+        self.conf = conf
+        self._resource = boto3.resource(
+            "dynamodb",
+            config=botocore.config.Config(region_name=region),
+            **self.conf)
+
+    def __getattr__(self, name):
+        return getattr(self._resource, name)
+
+
+class DynamoDBTable(threading.local):
+    def __init__(self, ddb_resource, *args, **kwargs):
+        # type: (DynamoDBResource, *Any, **Any) -> None
+        self._table = ddb_resource.Table(*args, **kwargs)
+
+    def __getattr__(self, name):
+        return getattr(self._table, name)
 
 
 class Message(object):
     """Create a Message table abstraction on top of a DynamoDB Table object"""
-    def __init__(self, table, metrics, max_ttl=MAX_EXPIRY):
-        # type: (Table, IMetrics) -> None
+    def __init__(self, tablename, metrics=None, boto_resource=None,
+                 max_ttl=MAX_EXPIRY):
+        # type: (str, IMetrics, DynamoDBResource, int) -> None
         """Create a new Message object
 
-        :param table: :class:`Table` object.
-        :param metrics: Metrics object that implements the
-                        :class:`autopush.metrics.IMetrics` interface.
+        :param tablename: name of the table.
+        :param metrics: unused
+        :param boto_resource: DynamoDBResource for thread
 
         """
-        self.table = table
-        self.metrics = metrics
+        self.tablename = tablename
         self._max_ttl = max_ttl
+        self.resource = boto_resource
+
+    def table(self, tablename=None):
```

Comment:
I don't think the table() method and all the associated changes for it are necessary now that the underlying 'table' object here would be the thread local wrapper

Reply:
We use `table()` to select between Message tables, so we'd still need to call DynamoDBTable() to fetch the correct reference to the table we'd want. 

For `Router` it makes more sense to drop `table()`, since there's only going to be one.

Reply:
What I mean is originally we had autopush.db.Message/Router with a thread safe '.table' boto2 Table object.

If we're turning '.table' into a thread safe DynamoDBTable wrapper, doesn't the need for a table() method (and the change to db.message_tables) go away? All this can continue working as before as long as the .table is thread safe (unless I'm missing something)